SpatialGDK/Source/SpatialGDKFunctionalTests/SpatialGDK/SpatialTestReplicationConditions/SpatialTestReplicationConditions.cpp

```diff
@@ -0,0 +1,598 @@
+// Copyright (c) Improbable Worlds Ltd, All Rights Reserved
+
+#include "SpatialTestReplicationConditions.h"
+
+#include "SpatialFunctionalTestFlowController.h"
+#include "SpatialGDKSettings.h"
+#include "TestReplicationConditionsActor.h"
+
+#include "Kismet/GameplayStatics.h"
+#include "Net/UnrealNetwork.h"
+
+/**
+ * This tests that Unreal replication conditions.
+ * This includes test for all replication conditions as defined in by ELifetimeCondition expect -
+ * COND_InitialOnly - tested in SpatialTestInitialOnly
+ * COND_InitialOrOwner - not supported in Spatial
+ * COND_ReplayOnly - not relevant to Spatial
+ *
+ * The test includes 2 Server and 2 Clients.
+ * The flow is as follows:
+ * - Setup:
+ *  - The test actors are spawned with both static and dynamic components.
+ *  - The test actors are possessed by Client1
+ * - Test:
+ *  - The test actors are their properties initialized with unique starting values.
+ *  - The non-auth server, owning client and non-owning client checks it can see the actors, components, and the values are correct.
+ *  - The test actors have their properties updated to new unique values.
+ *  - The non-auth server, owning client and non-owning client checks it can see the actors, components, and the values are correct.
+ * - Clean-up:
+ *	- The test actors are destroyed.
+ */
+ASpatialTestReplicationConditions::ASpatialTestReplicationConditions()
+	: Super()
+{
+	Author = "Mike";
+	Description = TEXT("Test Unreal Replication Conditions in a MultiServer Context");
+	static_assert(COND_Max == 16, TEXT("New replication condition added - add more tests!"));
+}
+
+void ASpatialTestReplicationConditions::GetLifetimeReplicatedProps(TArray<FLifetimeProperty>& OutLifetimeProps) const
+{
+	Super::GetLifetimeReplicatedProps(OutLifetimeProps);
+
+	DOREPLIFETIME(ThisClass, TestActor_Common);
+	DOREPLIFETIME(ThisClass, TestActor_CustomEnabled);
+	DOREPLIFETIME(ThisClass, TestActor_CustomDisabled);
+	DOREPLIFETIME(ThisClass, TestActor_AutonomousOnly);
+	DOREPLIFETIME(ThisClass, TestActor_PhysicsEnabled);
+	DOREPLIFETIME(ThisClass, TestActor_PhysicsDisabled);
+}
+
+void ASpatialTestReplicationConditions::PrepareTest()
+{
+	Super::PrepareTest();
+
+	bSpatialEnabled = GetDefault<UGeneralProjectSettings>()->UsesSpatialNetworking();
+	ReplicationStage = STAGE_InitialReplication;
```

Comment:
This should probably be set in an initial setup step instead, otherwise this test can't be rerun, right?
Maybe also the thing above and below unless you need to know them super-super early.